sdks/java/io/contextual-text-io/src/main/java/org/apache/beam/sdk/io/ContextualTextIO/ContextualTextIO.java

```diff
@@ -0,0 +1,554 @@
+      PCollection<LineContext> output = null;
+      if (getMatchConfiguration().getWatchInterval() == null && !getHintMatchesManyFiles()) {
+        output = input.apply("Read", org.apache.beam.sdk.io.Read.from(getSource()));
+      } else {
+        // All other cases go through FileIO + ReadFiles
+        output =
+            input
+                .apply(
+                    "Create filepattern", Create.ofProvider(getFilepattern(), StringUtf8Coder.of()))
+                .apply("Match All", FileIO.matchAll().withConfiguration(getMatchConfiguration()))
+                .apply(
+                    "Read Matches",
+                    FileIO.readMatches()
+                        .withCompression(getCompression())
+                        .withDirectoryTreatment(DirectoryTreatment.PROHIBIT))
+                .apply("Via ReadFiles", readFiles().withDelimiter(getDelimiter()));
+      }
+
+      // Output Contains LineContext Objects Without Correct Line Numbers
+      // The following operation assigns line numbers to all LineContext Objects
+
+      PCollection<KV<KV<String, Long>, Iterable<LineContext>>> groupedOutput =
+          output
+              .apply(
+                  "Convert LineContext to KV<KV<File,Range>, LineContext>",
+                  ParDo.of(
+                      new DoFn<LineContext, KV<KV<String, Long>, LineContext>>() {
+                        @ProcessElement
+                        public void processElement(
+                            @Element LineContext line,
+                            OutputReceiver<KV<KV<String, Long>, LineContext>> out) {
+                          out.output(
+                              KV.of(KV.of(line.getFile(), line.getRange().getRangeNum()), line));
+                        }
+                      }))
+              .apply("Apply GBK to PColl<KV<KV<File, Range>, LineCtx>>", GroupByKey.create());
+
+      PCollectionView<Map<KV<String, Long>, Long>> sizes =
+          groupedOutput
+              .apply(
+                  "KV<KV<File, Range>, Iter<LineCtx>> to KV<Range, Sizeof(Iter<LineCtx>)>",
+                  ParDo.of(
+                      new DoFn<
+                          KV<KV<String, Long>, Iterable<LineContext>>,
+                          KV<KV<String, Long>, Long>>() {
+                        @ProcessElement
+                        public void processElement(
+                            @Element KV<KV<String, Long>, Iterable<LineContext>> elem,
+                            OutputReceiver<KV<KV<String, Long>, Long>> out) {
+                          out.output(KV.of(elem.getKey(), (long) Iterables.size(elem.getValue())));
+                        }
+                      }))
+              .apply("Convert Sizes to PCollView", View.asMap());
+
+      // Get Pipeline to create a dummy PCollection with one element so that
+      // prefix sums can be computed in one pass
```

Comment:
The concept of prefix sums` may not be intuitive to a reader. If we start using 'chunk' terminology, perhaps we can use use `chunk offsets`/`chunk absolute offset` `offset of each chunk within the file`.